perses/analysis/free_energies.py

```diff
@@ -0,0 +1,242 @@
+import numpy as np
+import seaborn as sns
+from pymbar import BAR
+import matplotlib.pyplot as plt
+import seaborn
+import pickle
+from simtk.openmm import unit
+import bz2
+import json
+import tqdm
+from openmmtools.constants import kB
+import random
+import joblib
+from typing import Optional
+
+
+def _strip_outliers(w, n_devs=100):
+    w = [x for x in w if np.abs(x) < 10 ** 4]
+    mean = np.mean(w)
+    std = np.std(w)
+    good_w = [x for x in w if np.abs(x - mean) < n_devs * std]
+    return np.asarray(good_w)
+
+
+def _get_works(df, run, project, GEN=None):
+    works = df[(df["RUN"] == run)]
+
+    if GEN:
+        works = works[works["GEN"] == GEN]
+
+    f = works[works["PROJ"] == project].forward_work
+    r = works[works["PROJ"] == project].reverse_work
+    return f, r
+
+
+def free_energies(
+    details_file_path: str,
+    work_file_path: str,
+    complex_project: str,
+    solvent_project: str,
+    temperature_kelvin: float = 300.0,
+    n_bootstrap: int = 100,
+    show_plots: bool = False,
+    plot_file_format: str = str,
+    cache_dir: Optional[str] = None,
+):
+    r"""Compute free energies from a set of runs.
+
+    Parameters
+    ----------
+    details_file_path : str
+        Path to json file containing run metadata.
+    work_file_path : str
+        Path to bz2-compressed pickle file containing work values from simulation
+    complex_project : str
+        Project identifier (of the form "PROJXXXXX") of the FAH project for complex simulation
+    solvent_project : str
+        Project identifier (of the form "PROJXXXXX") of the FAH project for solvent simulation
+    temperature_kelvin : float
+    n_bootstrap : int, optional
+        Number of bootstrap steps used for BAR free energy estimation
+    show_plots : bool, optional
+        If true, block to display plots interactively (using `plt.show()`).
+        If false, save plots as images in current directory (for batch usage).
+    plot_file_format : str, optional
+        Image file format for saving plots. Accepts any file extension supported by `plt.savefig()`
+    cache_dir : str, optional
+        If given, local directory for caching BAR calculations. Results are cached by run, phase, and generation.
+        If None, no caching is performed.
+    """
+
+    with bz2.BZ2File(work_file_path, "r") as infile:
+        work = pickle.load(infile)
+
+    with open(details_file_path, "r") as f:
+        details = json.load(f)
+
+    work.replace([np.inf, -np.inf], np.nan)
```

Comment:
There's a context manager for `pandas` that helps with dropping infinities without needed to replace nothing (hence cheaper). See [here](https://stackoverflow.com/a/45746209).

Reply:
Cool!

Reply:
Switched to this in 55a091e.